chain/network/src/peer/peer_actor.rs

```diff
@@ -979,6 +691,48 @@ impl PeerActor {
         }
     }
 
+    fn receive_message(
+        &self,
+        ctx: &mut actix::Context<Self>,
+        conn: &connection::Connection,
+        msg: PeerMessage,
+    ) {
+        // This is a fancy way to clone the message iff event_sink is non-null.
+        // If you have a better idea on how to achieve that, feel free to improve this.
+        let message_processed_event = self
+            .network_state
+            .config
+            .event_sink
+            .delayed_push(|| Event::MessageProcessed(msg.clone()));
+        let was_requested = match &msg {
+            PeerMessage::Block(block) => {
+                let hash = *block.hash();
+                conn.chain_height.fetch_max(block.header().height(), Ordering::Relaxed);
+                self.tracker.lock().push_received(hash);
+                self.tracker.lock().has_request(&hash)
+            }
+            _ => false,
+        };
+        let clock = self.clock.clone();
+        let network_state = self.network_state.clone();
+        let peer_id = conn.peer_info.id.clone();
+        ctx.spawn(
+            wrap_future(async move {
+                network_state.receive_message(&clock, peer_id, msg, was_requested).await
```

Comment:
It looks a bit strange that this is routed via the global `network_state` rather than local `connection`.

Reply:
I want this PeerMessage handler to be agnostic to where the request came from (in particular the response should be sent outside of the handler). It is true that right now it is far from perfect, because even Client wants to get sometimes the peer_id of the sender for some reason. I'll move it back to PeerActor for now, and I'll think how to rearrange it later.